### PR TITLE
[AMDGPU] Do not emit arch dependent macros with unspecified cpu

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
Currently, the AMDGPU toolchain accepts not passing `-mcpu` as a means
to create a sort of "generic" IR. The resulting IR will not contain any
target dependent attributes and can then be inserted into another
program via `-mlink-builtin-bitcode` to inherit its attributes.

However, there are a handful of macros that can leak incorrect
information when compiling for an unspecified architecture. Currently,
things like the wavefront size will default to 64, which is actually
variable. We should not expose these macros unless it is known.

### CODE REVIEW

**Comment by jhuber6** [comment]
Rework of https://github.com/llvm/llvm-project/pull/79660 to handle old behavior of these being defined for the host.

File: clang/docs/HIPSupport.rst [hunk 1]
```diff
@@ -175,6 +175,8 @@ Predefined Macros
      - Defined when the GPU default stream is set to per-thread mode.
    * - ``HIP_API_PER_THREAD_DEFAULT_STREAM``
      - Alias to ``__HIP_API_PER_THREAD_DEFAULT_STREAM__``. Deprecated.
+   * - ``__AMDGCN_WAVEFRONT_SIZE__``
```
Comment (yxsamliu) [inline clang/docs/HIPSupport.rst:178]: this macro and other AMDGPU predefined macros are documented at https://clang.llvm.org/docs/AMDGPUSupport.html

we should not just mention the link and talk about their dependency on specific GPU arch and taking default value in host compilation.

File: clang/test/Preprocessor/predefined-arch-macros.c [hunk 2]
```diff
@@ -4294,13 +4294,20 @@
...
-// CHECK_AMDGCN: #define __HAS_FMAF__ 1
-// CHECK_AMDGCN: #define __HAS_FP64__ 1
-// CHECK_AMDGCN: #define __HAS_LDEXPF__ 1
+// CHECK_AMDGCN_803: #define __HAS_FMAF__ 1
+// CHECK_AMDGCN_803: #define __HAS_FP64__ 1
+// CHECK_AMDGCN_803: #define __HAS_LDEXPF__ 1
+// CHECK_AMDGCN_NONE-NOT: #define __HAS_FMAF__
+// CHECK_AMDGCN_NONE-NOT: #define __HAS_FP64__
+// CHECK_AMDGCN_NONE-NOT: #define __HAS_LDEXPF__
+// CHECK_AMDGCN_NONE-NOT: #define __AMDGCN_WAVEFRONT_SIZE__
```
Comment (yxsamliu) [inline clang/test/Preprocessor/predefined-arch-macros.c:4311]: we need a test for HIP host compilation by -x hip --offload-host-only

File: clang/docs/HIPSupport.rst (hunk 1 above)
Comment (arsenm) [reply clang/docs/HIPSupport.rst:178]: do these really need to be defined for the host?

Comment (jhuber6) [reply clang/docs/HIPSupport.rst:178]: No, it's explicitly wrong, but there's a lot of HIP headers that depend on it for some reason so we can't break it.

**Review by yxsamliu — APPROVED** [review]
LGTM. Thanks

**Comment by nico** [comment]
This seems to break tests: http://45.33.8.238/linux/129493/step_7.txt

Please take a look and revert for now if it takes a while to fix.

**Comment by jhuber6** [comment]
> This seems to break tests: http://45.33.8.238/linux/129493/step_7.txt
> 
> Please take a look and revert for now if it takes a while to fix.

Is it still broken? I pushed a fix because I'm pretty sure the problem was not passing `-nogpulib` `-nogpuinc` so the test runs on machines without ROCm. I always forget about that part because I have ROCm on my machine. It used to work silently before a patch last year or something that made `clang` failing make the test fail.

**Comment by nico** [comment]
This is with that change: http://45.33.8.238/linux/129493/step_7.txt

**Comment by nico** [comment]
i.e. it helped with   Clang :: Preprocessor/predefined-arch-macros.c but not with:

Failed Tests (2):
  Clang :: Driver/amdgpu-macros.cl
  Clang :: Driver/target-id-macros.cl

**Comment by jhuber6** [comment]
> i.e. it helped with Clang :: Preprocessor/predefined-arch-macros.c but not with:
> 
> Failed Tests (2): Clang :: Driver/amdgpu-macros.cl Clang :: Driver/target-id-macros.cl

Thanks, seeing it locally now. I'll try to fix it quick and revert if it's not working soon.

**Comment by jhuber6** [comment]
> i.e. it helped with Clang :: Preprocessor/predefined-arch-macros.c but not with:
> 
> Failed Tests (2): Clang :: Driver/amdgpu-macros.cl Clang :: Driver/target-id-macros.cl

Pushed a fix, `check-clang` passes on my machine now. Let me know if it's still broken.

**Comment by nico** [comment]
My bot is happy again, thanks.

Please run tests before committing :)

**Pull request merged**